### ISSUE TITLE
"IDE0004 C# Cast is redundant" is reported incorrectly for default

### ISSUE THREAD

**Issue by abatishchev** [dotnet/roslyn#49954]
**Version Used**: 

I'm sorry for dumb question but how to tell?

Environment:

```
VisualStudio.16.Release/16.8.3+30804.86
```

**Steps to Reproduce**:

```c#
protected bool? IsNewResource() =>
    Boolean.TryParse(GetHeader(HttpHeaders.NewResource), out var b) ? b : (bool?)default; 
```

**Expected Behavior**:

No issue is reported.

**Actual Behavior**:

Modified code:

```c#
protected bool? IsNewResource() =>
    Boolean.TryParse(GetHeader(HttpHeaders.NewResource), out var b) ? b : default as bool?;
```

doesn't compile:

>CS8716	There is no target type for the default literal.

**Comment by huoyaoyuan** [comment]
Simply removing the cast is valid in C# 9. The codefix should be aware of this.

**Comment by abatishchev** [comment]
I pasted the code that is being automatically generated 

Also having the cast as follows

```c#
protected bool? IsNewResource() =>
    Boolean.TryParse(GetHeader(HttpHeaders.NewResource), out var b) ? b : default;
```

returns `default(bool)` rather than `default(bool?)` causing the change in behavior and a bug: when header is missing the function returns a value while it shouldn't.

**Comment by abatishchev** [comment]
It should but it doesn't :) I pasted the code that is being automatically generated.

Today I tried it again and it did indeed generate proper code: just `default`, not `default as bool`. Wondering why.

But still, please see the testing code above, the resulting, modified code behaves differently.

**Comment by jinujoseph** [comment]
cc @CyrusNajmabadi

**Comment by abatishchev** [comment]
@CyrusNajmabadi:

Still an issue in VS 2022 (17.0.1):

![image](https://user-images.githubusercontent.com/351644/145149559-9969c774-4b2a-43a0-ba38-9483eee8752e.png)

Why it was closed?

**Comment by CyrusNajmabadi** [comment]
> Why it was closed?

Because it has been fixed in 17.1.  The build that https://github.com/dotnet/roslyn/pull/56932 went into.

**Comment by abatishchev** [comment]
Lovely!
How can I get the update? Is it available already? Would it be available with the next release of VS?

**Comment by CyrusNajmabadi** [comment]
17.1 Preview 1.1 is available here: https://visualstudio.microsoft.com/vs/preview/

Otherwise, this will be in the next non-preview update when 17.1 moves out of preview. :)
